src/main/java/duke/program/TaskManager.java

```diff
@@ -0,0 +1,404 @@
+            printIndividualTask(tasks.get(i), taskType, taskStatus, taskDescription);
+        }
+    }
+
+    public void addNewTodo(String rawDescription) throws EmptyDescriptionException {
+        String description = rawDescription.trim();
+        if (description.isEmpty()) {
+            throw new EmptyDescriptionException();
+        }
+        tasks.add(new ToDo(description));
+        taskCount++;
+    }
+
+    public void addNewDeadline(String descriptionAndTime) throws EmptyDescriptionException, InvalidFormatException {
+
+        String[] descriptionAndByTimeArray = descriptionAndTime.split(SEPARATOR_BY);
+        String description = descriptionAndByTimeArray[0].trim();
+
+        if (description.isEmpty()) {
+            throw new EmptyDescriptionException();
+        }
+        if (descriptionAndByTimeArray.length == 1) {
+            throw new InvalidFormatException();
+        }
+
+        String byDateTime = descriptionAndByTimeArray[1].trim();
+
+        tasks.add(new Deadline(description, byDateTime));
+        taskCount++;
+    }
+
+    public void addNewEvent(String descriptionAndTime) throws EmptyDescriptionException, InvalidFormatException {
+
+        String[] descriptionAndAtTimeArray = descriptionAndTime.split(SEPARATOR_AT);
+        String description = descriptionAndAtTimeArray[0].trim();
+
+        if (description.isEmpty()) {
+            throw new EmptyDescriptionException();
+        }
+        if (descriptionAndAtTimeArray.length == 1) {
+            throw new InvalidFormatException();
+        }
+
+        String startAndEndTime = descriptionAndAtTimeArray[1].trim();
+
+        tasks.add(new Event(description, startAndEndTime));
+        taskCount++;
+    }
+
+    public void printAddedTaskMessage(Task task) {
+
+        String taskType = task.getType();
+        String taskDescription = task.getDescription();
+
+        System.out.println("Got it. I've added this task:");
+        if (taskType.equals(TASK_TYPE_ICON_TODO)) {
+            System.out.printf("[%s][ ] %s%n", taskType, taskDescription);
+        } else if (taskType.equals(TASK_TYPE_ICON_DEADLINE)) {
+            String byTime = task.getByDateTime();
+            System.out.printf("[%s][ ] %s(by:%s)%n", taskType, taskDescription, byTime);
+        } else if (taskType.equals((TASK_TYPE_ICON_EVENT))) {
+            String atTime = task.getStartAndEndTime();
+            System.out.printf("[%s][ ] %s(at:%s)%n", taskType, taskDescription, atTime);
+        } else {
+            printGenericErrorMessage();
+        }
+
+        System.out.println("Now you have " + taskCount + " tasks in the list.");
+    }
+
+    public void markAsDone(ArrayList<Task> tasks, String[] lineArgs) throws InvalidIndexException, TaskIndexOutOfBoundsException {
+
+        if (lineArgs.length > 2) {
+            throw new InvalidIndexException();
+        }
+
+        int doneIndex = Integer.parseInt(lineArgs[1]) - 1;
+
+        if (doneIndex >= taskCount || doneIndex < 0) {
```

Comment:
It is not very clear what is the normal[ execution path](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent) is. You start off with catching an exception.